Translate renderToReadableStream (#670)
reactjs/es.react.dev: src/content/reference/react-dom/server/renderToReadableStream.md

}
```

Desde el cliente, llama a [`hydrateRoot`](/reference/react-dom/client/hydrateRoot) para hacer interactivo el HTML generado en el servidor.

[Consulta más ejemplos abajo.](#usage)

#### Parámetros {/*parameters*/}

> * `reactNode`: Nodo de React que quieras renderizar a HTML. Por ejemplo, un elemento JSX como `<App />`. Se presupone que representará el documento completo, por lo que el componente `App` debería renderizar la etiqueta `<html>`.

Reviewer: ```suggestion
* `reactNode`: Es un nodo de React que quieres renderizar en HTML. Por ejemplo, un elemento JSX como `<App />`. Se presupone que representará el documento completo, por lo que el componente `App` debería renderizar la etiqueta `<html>`.
```

* **opcional** `options`: Objeto con las opciones de streaming.
  * **opcional** `bootstrapScriptContent`: Si se especifica, esta cadena será colocada dentro de una etiqueta `<script>`.
  * **opcional** `bootstrapScripts`: Array de strings que representan las URL para que la etiqueta `<script>` las emita en la página. Utiliza esto para incluír el `<script>` que llama a [`hydrateRoot`.](/reference/react-dom/client/hydrateRoot) Omítelo si no planeas usar React en el cliente.
  * **opcional** `bootstrapModules`: Igual que `bootstrapScripts`, pero emite [`<script type="module">`](https://developer.mozilla.org/es/docs/Web/JavaScript/Guide/Modules) en su lugar.
  * **opcional** `identifierPrefix`: Prefijo (string) que React usa para los IDs generados mediante [`useId`.](/reference/react/useId) Útil para evitar conflictos cuando se usan múltiples raíces (roots) en la misma página. Debe ser el mismo prefijo que se le pasa a [`hydrateRoot`.](/reference/react-dom/client/hydrateRoot#parameters)

>   * **opcional** `namespaceURI`: Un string con la raíz del [namespace URI](https://developer.mozilla.org/es/docs/Web/API/Document/createElementNS#namespace_uris_válidos) para el stream. Por defecto se usa HTML. Usa `'http://www.w3.org/2000/svg'` para SVG o `'http://www.w3.org/1998/Math/MathML'` para MathML.

Reviewer: Kudos por encontrar la URL con traducción al castellano 🥇

>   * **opcional** `nonce`: Un string [`nonce`](http://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#nonce) que le permita al script la [`script-src` política de contenido seguro (CSP)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src).

Reviewer: ```suggestion
  * **opcional** `nonce`: Un string [`nonce`](http://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#nonce) que le permite al script la [`script-src` política de contenido seguro (CSP)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src).
```

>   * **opcional** `onError`: Retrollamada que se activa cuando ocurre un error en el servidor, sea [recuperable](#recovering-from-errors-outside-the-shell) o [no.](#recovering-from-errors-inside-the-shell) Por defecto, solo ejecuta `console.error`. Si lo sobreescribes para [registrar errores en el servidor,](#logging-crashes-on-the-server) asegúrate que sigues llamando a `console.error`. También puedes usarlo para [ajustar el código de estado](#setting-the-status-code) antes de que el shell sea emitido.

Reviewer: ```suggestion
  * **opcional** `onError`: Callback que se activa cuando ocurre un error en el servidor, sea [recuperable](#recovering-from-errors-outside-the-shell) o [no.](#recovering-from-errors-inside-the-shell) Por defecto, solo ejecuta `console.error`. Si lo sobreescribes para [registrar errores en el servidor,](#logging-crashes-on-the-server) asegúrate que sigas llamando a `console.error`. También puedes usarlo para [ajustar el código de estado](#setting-the-status-code) antes de que el shell sea emitido.
```

  * **opcional** `progressiveChunkSize`: El número de bytes en un chunk. [Leer más sobre el heurístico que se usa por defecto.](https://github.com/facebook/react/blob/14c2be8dac2d5482fda8a0906a31d239df8551fc/packages/react-server/src/ReactFizzServer.js#L210-L225)
  * **opcional** `signal`: Una [señal de aborto](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) que permite [abortar el renderizado en el servidor](#aborting-server-rendering) y el resto del renderizado en el cliente.

#### Devuelve {/*returns*/}

`renderToReadableStream` devuelve una promesa:

> - Si el [shell](#specifying-what-goes-into-the-shell) se renderizó satisfactoriamente, la promesa se resolverá a un [Readable Web Stream.](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream)

Reviewer: ```suggestion
- Si el [shell](#specifying-what-goes-into-the-shell) se renderizó satisfactoriamente, la promesa se devolverá a un [Readable Web Stream.](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream)
```

> - Si el renderizado del shell falla, la promesa será rechazada. [Usa esto para mostrar un shell mediante retrollamada.](#recovering-from-errors-inside-the-shell)

Reviewer: ```suggestion
- Si el renderizado del shell falla, la promesa será rechazada. [Usa esto para mostrar un shell mediante un callback.](#recovering-from-errors-inside-the-shell)
```

El stream devuelto tiene una propiedad adicional:

* `allReady`: Promesa que se resuelve cuando todo el renderizado se ha completado, incluyendo tanto el [shell](#specifying-what-goes-into-the-shell) como todo el [contenido](#streaming-more-content-as-it-loads) adicional. Puedes esperar a `await stream.allReady` antes de devolver una respuesta [para rastreadores y generación estática.](#waiting-for-all-content-to-load-for-crawlers-and-static-generation) Si usas esto, no obtendrás carga progresiva. El stream contendrá el HTML final.

---

## Uso {/*usage*/}

> ### Renderizar árbol de React como HTML a Readable Web Stream {/*rendering-a-react-tree-as-html-to-a-readable-web-stream*/}

Reviewer: ```suggestion
### Renderizar un árbol de React como HTML a un Readable Web Stream {/*rendering-a-react-tree-as-html-to-a-readable-web-stream*/}
```

}
```

React inyectará el [doctype](https://developer.mozilla.org/es/docs/Glossary/Doctype) y tus <CodeStep step={2}>etiquetas de `<script>` de arranque</CodeStep> dentro del HTML resultante del stream:

```html [[2, 5, "/main.js"]]
<!DOCTYPE html>
<html>
  <!-- ... HTML de tus componentes ... -->
</html>
<script src="/main.js" async=""></script>

>   ```

Reviewer: ```suggestion
```
```

> En el cliente, tu script de arranque debería [hidratar el `document` entero con una llamada a `hydrateRoot`:](/reference/react-dom/client/hydrateRoot#hydrating-an-entire-document)

Reviewer: ```suggestion
En el cliente, tu script inicial debería [hidratar el `document` entero con una llamada a `hydrateRoot`:](/reference/react-dom/client/hydrateRoot#hydrating-an-entire-document)
```

hydrateRoot(document, <App />);
```

Esto adjuntará detectores de eventos al HTML generado en el servidor, haciendo que este sea interactivo.

<DeepDive>

#### Leer recursos CSS y JS mediante su ruta a través del output {/*reading-css-and-js-asset-paths-from-the-build-output*/}

> Las URL finales de los recursos (como los archivos CSS y JS) suelen ser hasheadas después de la compilación. Por ejemplo, en lugar de `estilos.css` podrías tener `estilos.123456.css`. Hashear nombres de archivos estáticos garantiza que cada build distinta del mismo recurso tendrá un nombre de archivo diferente. Esto es útil porque te permite activar de forma segura el almacenamiento en caché a largo plazo para recursos estáticos: un archivo con un nombre característico nunca cambiaría su contenido.

Reviewer: ```suggestion
Las URL finales de los recursos (como los archivos CSS y JS) suelen ser hasheadas después de la compilación. Por ejemplo, en lugar de `estilos.css` podrías tener `estilos.123456.css`. El hasheo de nombres de archivos estáticos garantiza que cada versión distinta del mismo recurso tendrá un nombre de archivo diferente. Esto es útil porque te permite activar de forma segura el almacenamiento en caché a largo plazo para recursos estáticos: un archivo con un nombre característico nunca cambiaría su contenido.
```

> Sin embargo, si no conoces las URLs de los recursos hasta después de tener la build, no te será posible ponerlas en el código fuente. Por ejemplo, *hardcod*eas `"/styles.css"` en JSX, esto no funcionaría, puesto que es una URL relativa. Para mantenerlas fuera del código fuente, tu componente raíz puede leer el nombre real de un archivo a través de un mapa pasado como propiedad:

Reviewer: ```suggestion
Sin embargo, si no conoces las URLs de los recursos hasta después de tener el build, no te será posible ponerlas en el código fuente. Por ejemplo, si hardcodeas `"/styles.css"` en JSX, esto no funcionaría, puesto que es una URL relativa. Para mantenerlas fuera del código fuente, tu componente raíz puede leer el nombre real de un archivo a través de un mapa pasado como propiedad:
```

```js {1,6}
export default function App({ assetMap }) {
  return (
    <html>
      <head>
        <title>Mi app</title>
        <link rel="stylesheet" href={assetMap['estilos.css']}></link>
      </head>
      ...
    </html>
  );
}
```

En el servidor, renderiza `<App assetMap={assetMap} />` y pasa tu `assetMap` con las URLs de los recursos:

```js {1-5,8,9}

> // Necesitarás obtener este JSON a través de tus herramientas de compilación. Por ejemplo, leyéndolo desde laconsola de compilación

Reviewer: ```suggestion
// Necesitarás obtener este JSON a través de tus herramientas de compilación. Por ejemplo, leyéndolo desde la consola de compilación
```

```

En este ejemplo, React puede empezar a streamear el contenido mucho antes. Solo `ProfileLayout` y `ProfileCover` deben terminar de renderizar primero porque no están envueltos en ningún componente `<Suspense>`. Sin embargo, si `Sidebar`, `Friends`, o `Photos` necesitan cargar información, React enviará el HTML de `BigSpinner` en su lugar. Después, a medida que más información se vuelve disponible, se irá revelando el contenido hasta que toda la página haya cargado.

Cuando se streamea no es necesario esperar a que React en sí mismo cargue en el navegador o a que tu aplicación se vuelva interactiva. El contenido HTML irá siendo revelado progresivamente antes de que cualquier etiqueta `<script>` cargue.

[Lee más sobre cómo funciona el streaming de HTML.](https://github.com/reactwg/react-18/discussions/37)

<Note>

> **Solo las fuentes de información "Suspense-enabled" activarán el componente Suspense.** Se incluyen:

Reviewer: ```suggestion
**Solo las fuentes de información "Suspense-enabled" activarán el componente Suspense.** Estas incluyen:
```

- Obtención de datos con frameworks "Suspense-enabled" como [Relay](https://relay.dev/docs/guided-tour/rendering/loading-states/) y [Next.js](https://nextjs.org/docs/advanced-features/react-18)

> - Código de componentes lazy-loading de [`lazy`](/reference/react/lazy)

Reviewer: ```suggestion
- Código de componentes lazy-loading con [`lazy`](/reference/react/lazy)
```

Si el intento de cargar `Posts` desde el cliente *también* falla, React lanzará el error en el cliente. Igual que con el resto de errores lanzados durante el renderizado, el [límite de error más cercano](/reference/react/Component#static-getderivedstatefromerror) determinará cómo mostrar el error al usuario. En la práctica, esto significa que el usuario verá un símbolo de carga hasta que el error sea imposible de recuperar.

Si el intento de renderizado de `Posts` en el cliente se logra con éxito, el símbolo de carga recibido desde el servidor será reemplazado con el HTML obtenido como resultado. El usuario no sabrá que ocurrió un error en el servidor. Sin embargo, la función `onError` del servidor y la función [`onRecoverableError`](/reference/react-dom/client/hydrateRoot#hydrateroot) del cliente serán disparadas, por lo que podrás ser notificado/a del error.

---

### Establecer el código de estado {/*setting-the-status-code*/}

El streaming requiere de una compensación. Se pretende empezar el streaming lo antes posible de forma que el usuario vea el contenido lo antes posible también. No obstante, una vez se empieza el streaming, no es posible establecer un código de estado.

> Al [dividir la app](#specifying-what-goes-into-the-shell) en shell (por encima de los perímetros `<Suspense>`) y el resto de contenido, ya se ha solucionado una parte de este problema. Si el shell falla, el bloque `catch` te permitirá establecer el código de estado del error. De lo contrario, sabes que la app se podrá recuperar en el cliente, por lo que puedes enviar un "OK".

Reviewer: ```suggestion
Al [dividir la app](#specifying-what-goes-into-the-shell) en el shell (por encima de los perímetros `<Suspense>`) y el resto de contenido, ya se ha solucionado una parte de este problema. Si el shell falla, el bloque `catch` te permitirá establecer el código de estado del error. De lo contrario, sabes que la app se podrá recuperar en el cliente, por lo que puedes enviar un "OK".
```
